[GUI][Skin] Add new window dialogs to select video/audio/subtitles streams

CastagnaIT opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->
This PR introduce new dialog windows to select video/audio/subtitles streams, this superseed the old PR #16825, and also can allow a possible better integration of PR #22557 after that this one will be merged

This will also allow the opening of the dialog box to be mapped via shortcuts or remote which will further simplify the user experience

The implementation that can be fully customized in the skin XML by allowing skinners to make a nice customizable GUIs

On all dialogs:
- the star icon, means that is a stream marked as "default" from the metadata of media file or stream service (VOD, PVR, etc...), more details to default flag behaviour to comment https://github.com/xbmc/xbmc/pull/25762#issuecomment-2379051901
- the tick icon, mark the stream currently active

### Video dialog
First text line: codec and main characteristics + possible flags
Second text line: optional stream metadata description and/or angle view description
NOTE: i added "fps" info where before was missing, and was not possible to tell what kind of streams were
![immagine](https://github.com/user-attachments/assets/bd54a5ef-29ee-49d6-a748-349918f70cb5)

### Audio dialog
First line text: language + possible flags
Second line text: Codec description + channels
Third line text: optional stream metadata description
![immagine](https://github.com/user-attachments/assets/034a461c-95d8-4a3d-b572-1f77d1e903dc)

### Subtitle dialog
The first item "disable" is kept always on top, used to disable subtitles.
First text line: language + possible flags
Second text line: optional stream metadata description
![immagine](https://github.com/user-attachments/assets/83f39d92-87ce-4e8a-aa37-c33aa27a0ea6)


i have added a default items sort order to all streams types,
so that streams are listed with a kind of common sense,
where before i often saw a messy list (usually with InputStream Adaptive with services that provide many streams)
anyway this could be also improved in the future with new user settings preferences

NOTE: i leave untouched the existing GUI OSD -> Audio settings -> Stream list, that show always the textual list

Custom shortcut key example mapped to "o" key:
`<o>DialogSelectVideo</o>`
`<o>DialogSelectAudio</o>`
`<o>DialogSelectSubtitle</o>`

Most of keymaps that was using "next" stream has been replaced to use the new stream selection dialogs

New infolabel: `VideoPlayer.VideoStreamCount`

External subtitles: on the filenames are parsed also the flags "original" and "impaired", before missing

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
A method for selecting streams more easily and quickly has been requested several times over time by users,
main reasons are:
1) VOD streaming from services like am@zon/D+ etc... often have a lot of subtitle tracks such as 20~ tracks
the only way to change subititle stream e.g. from remote is to use "select next stream", this is a method that is no suitable for video services that provide many tracks.
2) users want to have a more direct way without constantly navigating through sub menus and after exit from GUI each time

solve https://forum.kodi.tv/showthread.php?tid=369842 (other improvements can be made after this, like filters/custom sorting)

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
VOD streaming

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->
A better user experience
Allow users to change video/audio/subtitles streams by using a window dialog that can be opened from GUI/remote/addons etc...

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

da-anda commented:
Thanks for picking up this feature again :+1: .

haven't checked the code, but just from looking a the screenshot I personally would miss a "disable subtitles/no subtitle" option or button (aka one dialog to rule them all). There should also be an indication if a subtitle only contains "forced" subs (which is a stream flag at least in MKV containers) and possibly also other indicators like "hearing impaired" etc. (via small icons maybe?).
Not sure if the numbering "2/3" etc is really needed - don't think it really adds any value to users. But maybe that's just me.

CastagnaIT commented:
@da-anda I agree on some points that im also was thinking,i published in advance also to get some comments as well
- i was also thinking to add a custom listitem entry to disable subs, would be very convenient instead to have a separate setting
- all stream flags are implemented, but i have kept as "text" as it was before, but skinners can customize as they wish with icons etc... imo i think it is clearer have text (forced, captions etc...) instead of a list of icons
- about numbering, i don't know i kept it as it was previously, but yes i think it actually doesnt have much sense especially if some kind of item sorting is applied, can be removed

ksooo commented on xbmc/video/guilib/VideoStreamSelectHelper.cpp:
```diff
@@ -0,0 +1,259 @@
+    std::string languageDesc;
+    g_LangCodeExpander.Lookup(info.language, languageDesc);
+
+    CFileItemPtr fileItem = std::make_shared<CFileItem>();
+    fileItem->SetLabel(info.name);
```
construction and setting the label can be done in one step:

`const auto fileItem{std::make_shared<CFileItem>(info.name)};`

ksooo commented on xbmc/video/guilib/VideoStreamSelectHelper.cpp:
```diff
@@ -0,0 +1,259 @@
+  // no streams - add a "None" entry
+  if (itemsToDisplay.IsEmpty())
+  {
+    CFileItemPtr fileItem = std::make_shared<CFileItem>();
+    fileItem->SetLabel(g_localizeStrings.Get(231));
```
See previous comment. Can be simplified.

ksooo commented on xbmc/video/guilib/VideoStreamSelectHelper.cpp:
```diff
@@ -0,0 +1,259 @@
+
+  if (dialog->IsConfirmed())
+  {
+    auto selectedItem = dialog->GetSelectedFileItem();
```
`const auto selectedItem`

ksooo commented on xbmc/video/guilib/VideoStreamSelectHelper.cpp:
```diff
@@ -0,0 +1,259 @@
+    if (!g_LangCodeExpander.Lookup(info.language, languageDesc))
+      languageDesc = g_localizeStrings.Get(13205); // Unknown
+
+    CFileItemPtr fileItem = std::make_shared<CFileItem>();
+    fileItem->SetLabel(languageDesc);
```
See previous comment. Can be simplified.

ksooo commented on xbmc/video/guilib/VideoStreamSelectHelper.cpp:
```diff
@@ -0,0 +1,259 @@
+  // no streams - add a "None" entry
+  if (itemsToDisplay.IsEmpty())
+  {
+    CFileItemPtr fileItem = std::make_shared<CFileItem>();
+    fileItem->SetLabel(g_localizeStrings.Get(231));
```
See previous comment. Can be simplified.

ksooo commented on xbmc/video/guilib/VideoStreamSelectHelper.cpp:
```diff
@@ -0,0 +1,259 @@
+
+  if (dialog->IsConfirmed())
+  {
+    auto selectedItem = dialog->GetSelectedFileItem();
```
`const auto selectedItem`

ksooo commented on xbmc/video/guilib/VideoStreamSelectHelper.cpp:
```diff
@@ -0,0 +1,259 @@
+    if (!g_LangCodeExpander.Lookup(info.language, languageDesc))
+      languageDesc = g_localizeStrings.Get(13205); // Unknown
+    
+    CFileItemPtr fileItem = std::make_shared<CFileItem>();
+    fileItem->SetLabel(languageDesc);
```
See previous comment. Can be simplified.

ksooo commented on xbmc/video/guilib/VideoStreamSelectHelper.cpp:
```diff
@@ -0,0 +1,259 @@
+  // no streams - add a "None" entry
+  if (itemsToDisplay.IsEmpty())
+  {
+    CFileItemPtr fileItem = std::make_shared<CFileItem>();
+    fileItem->SetLabel(g_localizeStrings.Get(231));
```
See previous comment. Can be simplified.

ksooo commented on xbmc/video/guilib/VideoStreamSelectHelper.cpp:
```diff
@@ -0,0 +1,259 @@
+
+  if (dialog->IsConfirmed())
+  {
+    auto selectedItem = dialog->GetSelectedFileItem();
```
`const auto selectedItem`

ksooo reviewed (commented):
Seeing lots of duplicated code in xbmc/video/guilib/VideoStreamSelectHelper.cpp - I suggest to consider refactoring to remove the duplicate code, please

CastagnaIT committed:
[GUI] Add videostreamcount infolabel

Hitcher commented on addons/skin.estuary/xml/Includes_SettingsDialog.xml:
```diff
@@ -34,9 +34,10 @@
 		<control type="button" id="22106">
 			<include>DialogSettingButton</include>
-			<label>$LOCALIZE[31115]</label>
+			<label>Change subtitles</label>
```
I assume this only temporary as it will need localisation.

CastagnaIT commented:
no one seems to look at the fact that the PR is DRAFT and WIP
i have not asked for code review only discussion about GUI appearance/usability referred to screenshot example

CastagnaIT commented on addons/skin.estuary/xml/Variables.xml:
```diff
@@ -965,4 +965,11 @@
 		<value condition="Integer.IsGreater(PVR.ClientCount,1)">$INFO[ListItem.PVRClientName,[COLOR grey]$LOCALIZE[31137]:[/COLOR] ,]$INFO[ListItem.PVRInstanceName, (,)]</value>
 	</variable>
+	<variable name="VideoSubsStreamDialogItemLabelVar">
+		<value condition="String.IsEqual(ListItem.Property(stream.isforced),true)+String.IsEqual(ListItem.Property(stream.ishearingimpaired),true)">$LOCALIZE[39106], $LOCALIZE[39107]</value>
```
@Hitcher about this var, example:
String.IsEqual(ListItem.Property(stream.isforced),true)
do you know if exists a bool comparison without use "string"?

jjd-uk replied:
All bools are listed in https://github.com/xbmc/xbmc/blob/master/xbmc/guilib/guiinfo/GUIInfoLabels.h and I don't see anything relevant hence probably why a ListItem.Property is being used.

ksooo commented:
> i have not asked for code review only discussion about GUI appearance/usability referred to screenshot example

Oops, I was of the opinion that feedback is always welcome. Will keep in mid that you do not want it unless you explicitly ask for it.

CastagnaIT commented:
yep, anyway its also missing pieces of code... probably i will finish it today

jjd-uk commented:
> discussion about GUI appearance/usability referred to screenshot example

In principle it looks good to me. However as already mention the 1/3, 2/3, 3/3 numbering is not needed in my opinion. I also would prefer having text for Default to be written in the label so you don't need to only rely on an icon (such as the star in the screenshot ) identify the default stream.

CastagnaIT commented:
> would prefer having text for Default to be written in the label

i dont see the reason to add it as text,
the "default" has no real useful meaning to understand the types of tracks like a subtitle to be choosen,
is just a courtesy to distiguish what kodi core chooses for the auto-selection of the track when the playback start,
is quite superfluous info, imo users should refer to kodi settings not to this "default" flag

da-anda commented:
thinking about it, marking the "default" does not really add any value. The user sees the "default" anyways, as it will be the currently selected/active stream when first opening the dialog. After the dialog had been opened/used once, seeing what KODI would have picked as default is no longer relevant, is it? In doubt we could simply move the "default" track to the top of the list.
Also, if what we mark as "default" does not reflect which stream has been tagged as default in the container (like MKV), tech savvy users could complain that it is mismatching and confusing. I'd probably use the spot with this icon to more clearly indicate the currently active stream (check mark?) - but that is just me.

CastagnaIT commented:
> Also, if what we mark as "default" does not reflect which stream has been tagged as default in the container (like MKV), tech savvy users could complain that it is mismatching and confusing

from what i know, the "default" flag is set only by video file metadata (such as MKV tracks) or from the VOD/Live/PVR streamining metadata ,and so on.... so the flag its not set or based from kodi settings.

from subtitle side: the default flag is used only if there are multiple identical subtitles with same language, that allow to auto-select one or the other one version. so most of the time is unused feature since kod setting dont have "media default" to "preferred subtitle language" setting

from audio side: this is useful only when  "media default" is set  to "preferred subtitle language" setting, in all other cases the default flag is useless.

from video side: allow kodi core to select the video to be choosen at playback start, but this is usefult only for "adaptive streaming" so multibitrate streaming and/or multicodec streaming, particular use cases

da-anda commented:
> from what i know, the "default" flag is set only by video file metadata (such as MKV tracks) or from the VOD/Live/PVR streamining metadata ,and so on.... so the flag its not set or based from kodi settings.

ok, so "default" would be what the media source/container defined, and not what KODI uses by default depending on user preferences (like in my case, I told KODI to use the English audio track over whatever is set as default). But still, I see little value of this information for end users. It is more a "nice to know" (for tech savvy users) that unnecessarily clutters the UI, than something a user would base any decision on. At least I would neither select an audio or subtitle stream based on a "default" flag. More important are things like language, audio codec and the various subtitle flags (forced, CC, ...)

CastagnaIT commented:
> ok, so "default" would be what the media source/container defined, and not what KODI uses by default depending on user preferences...

yep exactly, thats the reason why i said it's quite superfluous info
but although of little use if someone want to know the "default" tracks defined by media file/stream we must provide a little way to show it, so i choosen to show a small icon,
instead the most important info such as name, description, impaired types, etc... used to understand the tracks, are clearly visible as text

CastagnaIT committed:
[VideoPlayer] Decouple codec description from codec name

Our custom codec descriptions should not be mixed with codec name defined by ffmpeg
this will allow also skins to get the real codec name

CastagnaIT committed:
[VideoPlayer] Copy codecName to SubtitleStreamInfo

CastagnaIT commented:
ok, the hunger games begin, i think now PR is ready to receive a first review
i updated also PR description with new details screenshots

CastagnaIT committed:
[keymaps] Use video/audio/subtitle selection dialogs

ksooo commented on xbmc/cores/VideoPlayer/Interface/StreamInfo.h:
```diff
@@ -73,6 +73,8 @@ struct VideoStreamInfo : StreamInfo
   std::string stereoMode;
   int angles = 0;
   StreamHdrType hdrType = StreamHdrType::HDR_TYPE_NONE;
+  int fpsRate{0};
+  int fpsScale{0};
```
Can fps be negative? Maybe `unsigned int`is a better fit.

CastagnaIT replied:
done

ksooo commented on xbmc/cores/VideoPlayer/VideoPlayer.h:
```diff
@@ -197,6 +197,8 @@ struct SelectionStream
   std::string stereo_mode;
   float aspect_ratio = 0.0f;
   StreamHdrType hdrType = StreamHdrType::HDR_TYPE_NONE;
+  int fpsScale{0};
+  int fpsRate{0};
```
Can fps be negative? Maybe `unsigned int` is a better fit.

ksooo commented on xbmc/video/PlayerController.cpp:
```diff
@@ -224,19 +224,23 @@ bool CPlayerController::OnAction(const CAction &action)
-          aud = StringUtils::Format("{} - {}", lan, info.name);
+
+        std::string textInfo = lan;
+        if (!info.name.empty())
+          textInfo += " - " + info.name;
+        if (!info.codecDesc.empty())
+          textInfo += " (" + info.codecDesc + ")";
```
Maybe the format of the info should be localizable instead of hard coded? In other languages the dash (-) might not be the preferred delimiter for example.

CastagnaIT replied:
I just adapted the existing code that was already hardcoded with dash char,
there are many parts of code, not only this one, that could be localizable, but I don't adjust them on this PR, its already big

ksooo replied:
Not sure this is a wise discussion. After your PR got merged nobody will clean it up i fear.

CastagnaIT replied:
this code (and also the others "next"/"cycle" behaviours imo must be full reworked,
because now will result that stream dialogs have applied a sort order, while these do not, and so users "scroll" unordered streams

so the idea is use the new helper file added here to get an ordered list of streams (done with SubtitleStreamInfo, etc...) and so after apply the "next"/"cycle" behaviour, along with this changes can be add new localizables id's

if you really want I will do here, but it will result in very mess PR
since all these changes that are not few, are not directly related to dialogs implementations

CastagnaIT replied:
in my previous message i said "many parts of code",
because there are also other situations,
for example on skins xml, the text on `$INFO` is not localized
example:
https://github.com/xbmc/xbmc/blob/12ac3e2085665ae0f23fcb719e0aa67cbd433352/addons/skin.estuary/xml/DialogPlayerProcessInfo.xml#L240-L249
and same thing on the xml code that i add on this PR

unless i am not aware of it,
suggest that kodi core dont provide a kind of `$INFOLOC` "localizable" where you can inject the values to be added to the string as arguments directly from the xml,
this forces to apply localizations via source code, and does not allow provide clean values to skins, for example you have:
`Player.Process(audiochannels)` that provide the n° of channels
now to localize it, you forced to localize value as string from source code and will result in a string e.g.:
"channels: 4"
but in this way skins cant have a clean value of "4"
where they can manage as they wish

ksooo commented on xbmc/video/dialogs/GUIDialogAudioSettings.cpp:
```diff
@@ -346,31 +346,34 @@ void CGUIDialogAudioSettings::AudioStreamsOptionFiller(const SettingConstPtr& se
 
-    if (info.name.length() == 0)
-      info.name = strUnknown;
+    std::string textInfo = strLanguage;
+    if (!info.name.empty())
+      textInfo += " - " + info.name;
+
+    textInfo += " (";
+    if (!info.codecDesc.empty())
+      textInfo += info.codecDesc + ", ";
```
Maybe the format of the info should be localizable instead of hard coded? In other languages the dash (-) or comma (,) might not be the preferred delimiter for example.

CastagnaIT replied:
as above, i dont localize re-adapted old code on this PR

ksooo commented on xbmc/video/dialogs/GUIDialogAudioSettings.cpp:
```diff
@@ -346,31 +346,34 @@ void CGUIDialogAudioSettings::AudioStreamsOptionFiller(const SettingConstPtr& se
-    strItem += StringUtils::Format(" ({}/{})", i + 1, audioStreamCount);
-    list.emplace_back(strItem, i);
+    textInfo += FormatFlags(info.flags);
+    textInfo += StringUtils::Format(" ({}/{})", i + 1, audioStreamCount);
```
Strictly speaking, same localization problem as above.

CastagnaIT replied:
as above, i dont localize re-adapted old code on this PR

ksooo commented on addons/skin.estuary/language/resource.language.en_gb/strings.po:
```diff
@@ -502,7 +502,12 @@ msgctxt "#31107"
-#empty strings from id 31108 to 31109
+#empty strings from id 31108 to 31108
+
+#: /xml/VideoOSD.xml
```
We use the fully qualified name here => addons/skin.estuary/xml...
I know this is not used correctly at other places in this file, but for new strings we should do it right,

ksooo commented on addons/skin.estuary/language/resource.language.en_gb/strings.po:
```diff
@@ -516,7 +521,7 @@ msgstr ""
 
 #: /xml/VideoOSD.xml
 msgctxt "#31112"
-msgid "Toggle audio stream"
+msgid "Audio streams"
```
I consider this change risky as you never know whether other skins also use this string for whatever. With this change you might "break" them. I would introduce a new string instead of changing the existing string.

CastagnaIT replied:
with each new version of kodi skinner they have to adjust their thirdy party skins,
so i dont see real reason to keep obsolete texts that are no longer used on Kodi...
there are already other unused strings on files that just cause a mess by occupying id's

scott967 replied:
Have to admit I ran out of string space in 31000 block on aeon mq5 mods, so always am looking to the Kodi strings to "borrow" from.  Maybe not so much for something like "toggle audio stream" that is kind of use specific,  but "Subtitle" -> "Subtitle stream" could be considered to keep (just checked and I don't use it).  Note this also potentially affects any skinned addon windows.

ksooo replied:
Then you need to document this breaking changes for the skinners in the forum. We have a thread for that there.

CastagnaIT replied:
yes i will add to https://forum.kodi.tv/showthread.php?tid=377306

ksooo commented on addons/skin.estuary/language/resource.language.en_gb/strings.po:
```diff
@@ -531,7 +536,7 @@ msgstr ""
 
 #: /xml/VideoOSD.xml
 msgctxt "#31115"
-msgid "Toggle subtitle"
+msgid "Subtitles streams"
```
I consider this change risky as you never know whether other skins also use this string for whatever. With this change you might "break" them. I would introduce a new string instead of changing the existing string.

CastagnaIT replied:
replied above

ksooo commented on addons/skin.estuary/language/resource.language.en_gb/strings.po:
```diff
@@ -924,3 +929,9 @@ msgstr ""
 msgid "System"
 msgstr ""
+
+#: /xml/Includes_DialogSelect.xml
```
We use the fully qualified name here => addons/skin.estuary/xml...
I know this is not used correctly at other places in this file, but for new strings we should do it right.

CastagnaIT replied:
on kodi language files yes but i have never seen this on skin language files,
its a new rule?

ksooo replied:
If you search for addons/skin.estuary/ in this file you will get many hits.

ksooo replied:
Oh wait, you are right. This is strings.po of the estuary skin, not the global strings.po. Sorry.

ksooo commented on xbmc/video/guilib/VideoStreamSelectHelper.cpp:
```diff
@@ -0,0 +1,423 @@
+constexpr int FILEITEM_INDEX_NONE = -1; // FileItem index value for "none" stream
+
+// \brief Make a FileItem entry with "Disable" label, allow to disable a stream
+CFileItemPtr MakeFileItemDisable(const bool isSelected)
```
I suggest not to use the `typedef`, but the real type which is `std::shared_ptr<CFileItem>`. This is what we do with new code. Please change this here, and everywhere else in this file.

`const`makes no sense for primitive types when used as stack parameters. Remove the `const`, please.

CastagnaIT replied:
i usually add const to arguments like this isSelected because make in clear that the value will be not modified by the method code
if you remove the const the value can be modified by method

ksooo replied:
which for by value parameters is pointless. The caller simply does not need to know that the implementation does not modify the stack param.

ksooo replied:
https://stackoverflow.com/questions/8714250/isnt-const-redundant-when-passing-by-value

CastagnaIT replied:
it's not about caller, it's more about readability,
in this case the code is small and can also be removed,
but if you have a method with very long code, clearly state that the variable will not be modified by the method code, as described on the linked url

ksooo replied:
One more time: the const is not part of the contract. It means nothing to the *caller* whether it is present or not. You describe the contract with the method declaration , not implementation details. 

BUT, making the param const enforces the *implementation* to make a copy of the stack parameter which is already a copy in case the implementation wants to manipulate the parameters value. Thus you are taking the importation the chance to optimize the code for no reason.

CastagnaIT replied:
now i understood, i will do it

ksooo commented on xbmc/video/guilib/VideoStreamSelectHelper.cpp:
```diff
@@ -0,0 +1,423 @@
+// \brief Make a FileItem entry with "Disable" label, allow to disable a stream
+CFileItemPtr MakeFileItemDisable(const bool isSelected)
+{
+  CFileItemPtr fileItem = std::make_shared<CFileItem>(g_localizeStrings.Get(24021));
```
`const auto fileItem{std::make_shared...`- shorter and more readable and `const` - here and at other places in this file.

CastagnaIT replied:
here i cant use "const auto fileItem"
it will result that return std::shared_ptr<const CFileItem>
and dont work with CFileItemList uses

ksooo replied:
Then you can make it work by not using `auto` but `const std::shared_ptr<CFileItem> fileitem`.

CastagnaIT replied:
oh ok seem works also auto, i just didn't do the same with the method return value

ksooo commented on xbmc/video/guilib/VideoStreamSelectHelper.cpp:
```diff
@@ -0,0 +1,423 @@
+}
+
+std::shared_ptr<const CFileItem> OpenSelectDialog(CGUIDialogSelect& dialog,
+                                                  const int headingId,
```
Remove `const`.

CastagnaIT replied:
i remove but i want to point out that you have introduced similar code on:
VideoVersionHelper.cpp ChooseVideo method

ksooo replied:
Really? Then this was clearly a mistake on my end.

ksooo replied:
Look closer at CVideoVersionHelper -> ChooseVideo's paramters. I cannot see `const int headingId` there, only ìnt headingId`.

CastagnaIT replied:
but you are talking of const of const int headingId? because since you havent specified on first message i was thinking of  const CFileItem....

ksooo replied:
Yes, I’m talking about the const for param headingId.

ksooo commented on xbmc/video/guilib/VideoStreamSelectHelper.cpp:
```diff
@@ -0,0 +1,423 @@
+
+std::string CovertFpsToString(float value)
+{
+  if (value == 0)
```
`value == 0.0`- it's a float, you should make that clear.

garbear replied:
`value == 0.0f` even better

ksooo replied:
That’s what i meant. :-)

ksooo commented on xbmc/video/guilib/VideoStreamSelectHelper.cpp:
```diff
@@ -0,0 +1,423 @@
+  if (itemsToDisplay.IsEmpty())
+    itemsToDisplay.Add(MakeFileItemNone());
+
+  const auto selectedItem = OpenSelectDialog(*dialog, 38031, itemsToDisplay);
```
I would create the dialog instance inside `OpenSelectDialog`, not here. So, you safe the parameter. You do nothing with `dialog` here except to create the instance. Same for the other places where you create dialogs in this file.

ksooo reviewed (commented):
"Hunger games", episode 1. :-)

CastagnaIT commented on addons/skin.estuary/language/resource.language.en_gb/strings.po:
```diff
@@ -924,3 +929,9 @@ msgstr ""
+#: /xml/Includes_DialogSelect.xml
+#. Audio channels
+msgctxt "#31612"
+msgid "channels"
```
is not strictly necessary, depends on preferences
![immagine](https://github.com/user-attachments/assets/58252dc4-4add-4018-889e-544bff2d273d)

CastagnaIT committed:
[VideoPlayer][StreamInfo] Add fps info to VideoStreamInfo

Hitcher commented:
Off topic but I'm curious as to whether `stream.codecdesc` could be used to make a new info label for skinners to use to display the correct channel layout instead of just guessing it, as we currently do based on the number of channels?
See: [Request -  Audio channel layout info label](https://forum.kodi.tv/showthread.php?tid=377319) for more information.
Thanks.

CastagnaIT commented:
> Off topic but I'm curious as to whether `stream.codecdesc` could be used to make a new info label for skinners to use to display the correct channel layout instead of just guessing it, as we currently do based on the number of channels? See: [Request - Audio channel layout info label](https://forum.kodi.tv/showthread.php?tid=377319) for more information. Thanks.

nope `stream.codecdesc` for audio case, its a our custom text representation of the codec fourcc and codec profile, has nothing to do with channels

in the example screenshot above, you see 
![immagine](https://github.com/user-attachments/assets/4a3d29dd-84a7-4927-8565-8173fbb40390)
but the "5.1(side)" if a text addition that come from the MKV metadata (i will add a comma to make separate texts)
only the initial "AC3" text come from kodi core, instead the "channels: 6" its the same value that you are talking on the forum
this is used everywhere in kodi core suggest also in the database, seems to require different changes

CastagnaIT commented:
changes made, with only exceptions review comments not "closed"

ksooo commented on xbmc/video/guilib/VideoStreamSelectHelper.cpp:
```diff
@@ -0,0 +1,478 @@
+    fileItem->SetProperty("stream.language", languageDesc);
+
+    fileItem->SetProperty("stream.resolution",
+                          std::to_string(info.width) + "x" + std::to_string(info.height));
```
This is a good example for problems with hardcoding strings that should be localizable. The lower case letter x is not the best choice here, there is a unicode value that fits better, at least for my language. => U+2715 (10005) "times"

CastagnaIT replied:
you cant use U+2715 since its not always present on the font sets,
also the standard "x" char is used everywhere on kodi as on resolution lists, system info etc...

ksooo replied:
> also the standard "x" char is used everywhere on kodi as on resolution lists, system info etc...

That does not make it right. 

It’s a pity and not a serious issue in this case, but localization seems to be something not many of us want to spend time for improvements. 

Anyways, nothing we should consider a blocker for this PR.

CastagnaIT replied:
anyway it would still be a half done thing, should be found a solution for all use cases
the problems are also on skin side, e.g.
```
<label>$INFO[Player.Process(videowidth),[COLOR button_focus]$LOCALIZE[38031]:[/COLOR] ,x]$INFO[Player.Process(videoheight)]$INFO[Player.Process(videoscantype)]$INFO[Player.Process(videodar),$COMMA , AR]$INFO[Player.Process(videofps),$COMMA , FPS]</label>
```
these are not localizable

should be provided a way to allow skins to apply localizations on id's that use formatting placeholders, e.g.
i can create something like e.g.
```
msgctxt "#50000"
msgid "{0:d}x{1:d}"
msgstr ""
```
but if you use $LOCALIZE[50000] how you can send the Player.Process(videowidth) Player.Process(videoheight) values from xml?
maybe extending LOCALIZE to accept parameters
$LOCALIZE[50000, Player.Process(videowidth), Player.Process(videoheight), ... ]
that will format the string id with provided infolabels, the bigger problem are the value type conversions from string (CGUIInfoManager/infolabels seem to convert all to strings)

ksooo reviewed (commented):
Hunger games, episode 2

CastagnaIT commented on xbmc/cores/VideoPlayer/Interface/StreamInfo.h:
```diff
@@ -44,7 +44,8 @@ struct StreamInfo
   std::string language;
   std::string name;
-  std::string codecName;
+  std::string codecName; // Codec name (name definition from ffmpeg)
+  std::string codecDesc; // Codec description
```
i remember well the first few times i tried to understand how this code worked and it was not clear what "codec" was referred, so i thought it would be useful to have a descriptive comment on the variable since i made a small change here
i dont want touch tons of code to rename a variable that dont concern this PR,
same for the below `SelectionStream` that i add same comments,
i will remove all comments that you dont like

CastagnaIT replied:
removed

CastagnaIT commented on xbmc/addons/kodi-dev-kit/include/kodi/c-api/gui/input/action_ids.h:
```diff
@@ -677,6 +677,15 @@ enum ADDON_ACTION
   /// @brief <b>`267`</b>: Tempo decrease in current file played. global action, can be used anywhere
   ADDON_ACTION_PLAYER_DECREASE_TEMPO = 267,
 
+  /// @brief <b>`270 `</b>: Open the dialog window to select a video stream
+  ADDON_ACTION_DIALOG_SELECT_VIDEO = 270,
+
+  /// @brief <b>`271 `</b>: Open the dialog window to select a audio stream
+  ADDON_ACTION_DIALOG_SELECT_AUDIO = 271,
+
+  /// @brief <b>`272 `</b>: Open the dialog window to select a subtitle stream
+  ADDON_ACTION_DIALOG_SELECT_SUBTITLE = 272,
+
```
yeah seem so

CastagnaIT commented on xbmc/video/dialogs/GUIDialogAudioSettings.cpp:
```diff
@@ -346,31 +346,34 @@ void CGUIDialogAudioSettings::AudioStreamsOptionFiller(const SettingConstPtr& se
+  const int audioStreamCount = appPlayer->GetAudioStreamCount();
 
-  std::string strFormat = "{:s} - {:s} - {:d} " + g_localizeStrings.Get(10127);
+  std::string strChannels = g_localizeStrings.Get(10127);
```
renamed only my new variable

CastagnaIT commented on xbmc/video/guilib/VideoStreamSelectHelper.cpp:
```diff
@@ -0,0 +1,478 @@
+    info.isDefault = info.flags & StreamFlags::FLAG_DEFAULT;
+    info.isForced = info.flags & StreamFlags::FLAG_FORCED;
+    info.isHearingImpaired = info.flags & StreamFlags::FLAG_HEARING_IMPAIRED;
+    info.isVisualImpaired = info.flags & StreamFlags::FLAG_VISUAL_IMPAIRED;
```
i reworked code in better way, also changed "index" names to "id" since i realized that is called "id" also on others kodi core parts, removed also the needs of "pairs"

CastagnaIT commented:
changes pushed

ksooo commented on xbmc/video/guilib/VideoStreamSelectHelper.cpp:
```diff
@@ -0,0 +1,480 @@
+    isOriginal = info.flags & StreamFlags::FLAG_ORIGINAL;
+  }
+
+  int streamId;
+  std::string languageDesc;
+  bool isDefault;
+  bool isForced;
+  bool isHearingImpaired;
+  bool isVisualImpaired;
+  bool isOriginal;
```
I would init the bools and the int here, for safety.

CastagnaIT replied:
done

ksooo commented on xbmc/video/guilib/VideoStreamSelectHelper.cpp:
```diff
@@ -0,0 +1,480 @@
+    isVisualImpaired = info.flags & StreamFlags::FLAG_VISUAL_IMPAIRED;
+  }
+
+  int streamId;
+  std::string languageDesc;
+  bool isDefault;
+  bool isForced;
+  bool isHearingImpaired;
+  bool isVisualImpaired;
```
I would init the bools and the int here, for safety.

ksooo reviewed (commented):
Hunger games, episode 3. :-)

scott967 commented on xbmc/addons/kodi-dev-kit/include/kodi/versions.h:
```diff
@@ -49,8 +49,8 @@
-#define ADDON_GLOBAL_VERSION_GUI                      "5.15.0"
-#define ADDON_GLOBAL_VERSION_GUI_MIN                  "5.15.0"
+#define ADDON_GLOBAL_VERSION_GUI                      "5.16.0"
+#define ADDON_GLOBAL_VERSION_GUI_MIN                  "5.16.0"
```
Skins are already at 5.17.0?

CastagnaIT replied:
done
i forgot how it worked i know with binary addon i used to change both

CastagnaIT committed:
[addons][gui] increase GUI API version to 5.15.1

CastagnaIT commented:
jenkins build this please

CastagnaIT commented:
maybe @Hitcher or other skinners can review xml parts?

garbear commented on xbmc/video/guilib/VideoStreamSelectHelper.cpp:
```diff
@@ -0,0 +1,480 @@
+  }
+
+  auto& components = CServiceBroker::GetAppComponents();
+  auto appPlayer = components.GetComponent<CApplicationPlayer>();
```
Do we need to check for !appPlayer, or is appPlayer always assumed in the context of video code?

CastagnaIT replied:
by looking at `components.GetComponent<CApplicationPlayer>()` uses in the source code is never used check the pointer

jjd-uk commented:
> maybe @Hitcher or other skinners can review xml parts?

I plan to take a look at the weekend.

I see there's no response yet to:

> since now the main OSD menu contains the new buttons to show a/v/s stream list dialogs the old submenu " textual stream list" (on each separate setting) imo is no longer useful, is double thing I would eliminate them, but I would like to know your opinions, delete/kept/change them, other?

Personally I believe these should be additional options for the skinners to give maximum flexibility to the skinner on how they want to lay out access to these stream selections. Therefore access to the stream selections should still be retained within the Audio/Video/Subtitle settings dialogs as well in my view. However perhpas it would make sense for consistancy if the options in the Audio/Video/Subtitle settings dialogs could call these new dialogs instead of the old textual stream lists, if that can be done and is not too difficult.

For Estuary I'd like to see in a follow up PR where these 3 options moved to a new stream selection menu on the OSD like in https://github.com/xbmc/xbmc/pull/22557

CastagnaIT commented:
> For Estuary I'd like to see in a follow up PR where these 3 options moved to a new stream selection menu on the OSD like in https://github.com/xbmc/xbmc/pull/22557

yep the idea is allow the dev of that PR to rebase and align it to use new dialogs

> Personally I believe these should be additional options for the skinners to give maximum flexibility to the skinner on how they want to lay out access to these stream selections. Therefore access to the stream selections should still be retained within the Audio/Video/Subtitle settings dialogs as well in my view. However perhpas it would make sense for consistancy if the options in the Audio/Video/Subtitle settings dialogs could call these new dialogs instead of the old textual stream lists, if that can be done and is not too difficult.

for me its ok, i can see to do it

CastagnaIT commented:
>  However perhpas it would make sense for consistancy if the options in the Audio/Video/Subtitle settings dialogs could call these new dialogs instead of the old textual stream lists, if that can be done and is not too difficult.

I investigated, but it depends on what we want to visualize, and that could also be very challenging,
i think it's too much for this PR seems to require too many changes,
mainly because the a/v/s settings dialogs are not xml's, but hardcoded GUI dialogs via source code
and currently from what i see are missing parts of GUI implementations manageable via code...

the current menu that show audio/video/subs stream lists are a "button setting list", the list and the text shown is hardcoded
i mean e.g. this
![immagine](https://github.com/user-attachments/assets/b6d2f30c-e865-4ca6-918f-37c9dab48597)
it could be replaced with a "button setting"
but a button setting currently is not implemented to manage the label 2 (text on the right) via code, its full missing the implementation.
I tried add the label2 support, but, there are broader problems because currently the “click” callback system method (OnSettingAction) have read-only object, and so dont allow to update a custom text to the setting button control, and nor to refresh the GUI control text on screen, at least i couldn't find any easy way

even thinking about using infolabels to avoid manually update label2 is too limiting, currently there are no infolabels with the necessary information

imo possible simple solutions are:
- keep as is
- replace it with a simple button, without stream info text (on the right - label2)
- replace it with a text menu line (not clickable) only to show relevant info about current selected stream (not sure about feasibility/difficulty, sound better separate PR's, i havent found a way to do it by myself)
- try add miss GUI implementations to manage/update label2 on button menu, hard and require much more time, idk if im able to do it, and i have also other tasks atm
- other hints?

Hitcher commented:
Tested this morning and everything works as expected.

enen92 reviewed (commented):
Is changing `IPlayerSubtitleCapabilities` to an enum class something you'd be willing to do on the scope of this PR? Otherwise ignore

CastagnaIT committed:
[Players][cleanup] Changed audio/sub capabilities to enum class

CastagnaIT commented:
> Is changing `IPlayerSubtitleCapabilities` to an enum class something you'd be willing to do on the scope of this PR? Otherwise ignore

ok cleanup done

jjd-uk commented:
If @Hitcher found no problem on the skin side then I expect everything to be ok. I wasn't feeling well over weekend so didn't get to testing current state. I will find the time later today to check skin changes for myself.

CastagnaIT commented:
i made a small change, i have add support to "original" flag to subtitles, this was missing from old kodi versions so an old bug (also on textual subs list menu) found while i was reading #25617, I didn't even realize it...

CastagnaIT commented:
i found a problem with external subtitles
in the old textual stream selection all external sub tracks was forced placed all on top,
but now are listed in alphabetical order as all others e.g.
![immagine](https://github.com/user-attachments/assets/2db3e667-8311-4f53-94d5-cf8c9cbccaeb)
i will investigate

CastagnaIT committed:
[Utils] Add missing original, imparied filename parsing

CastagnaIT committed:
[GUI][Skin] Video/Audio/Subtitle stream selection dialog

CastagnaIT commented:
fixed, if someone want give a second code look
to fix it i add isExternal to SubtitleStreamInfo

add also an additional commit 193fc6a1e89db1fc86cde682753341040f4dc491
to add missing parsing original, impaired, that i used to test sort order

i think now its ready

CastagnaIT commented:
if no more requests, someone can approve?

basilgello reviewed (approved):
Thanks for great work!

CastagnaIT commented:
added changes to forum skinning engine thread https://forum.kodi.tv/showthread.php?tid=377306&pid=3212851#pid3212851

jenkins build this please

mackal commented:
`xbmc/cores/VideoPlayer/Interface/StreamInfo.h:79:3: error: ‘uint32_t’ does not name a type` missing include with GCC 14.

Hitcher commented on addons/skin.estuary/xml/Variables.xml:
@CastagnaIT I think these 3 variables can be reduced to one with the same results.

```
	<variable name="VideoStreamDialogItemLabelVar">
		<value condition="String.IsEqual(ListItem.Property(stream.isoriginal),true) + String.IsEqual(ListItem.Property(stream.isforced),true) + String.IsEqual(ListItem.Property(stream.ishearingimpaired),true)">$LOCALIZE[39111], $LOCALIZE[39106], $LOCALIZE[39107]</value>
		<value condition="String.IsEqual(ListItem.Property(stream.isoriginal),true) + String.IsEqual(ListItem.Property(stream.isforced),true) + String.IsEqual(ListItem.Property(stream.isvisualimpaired),true)">$LOCALIZE[39111], $LOCALIZE[39106], $LOCALIZE[39108]</value>
		<value condition="String.IsEqual(ListItem.Property(stream.isoriginal),true) + ListItem.Property(stream.isforced),true)">$LOCALIZE[39111], $LOCALIZE[39106]</value>
		<value condition="String.IsEqual(ListItem.Property(stream.isforced),true) + String.IsEqual(ListItem.Property(stream.ishearingimpaired),true)">$LOCALIZE[39106], $LOCALIZE[39107]</value>
		<value condition="String.IsEqual(ListItem.Property(stream.isforced),true) + String.IsEqual(ListItem.Property(stream.isvisualimpaired),true)">$LOCALIZE[39106], $LOCALIZE[39108]</value>
		<value condition="String.IsEqual(ListItem.Property(stream.ishearingimpaired),true)">$LOCALIZE[39107]</value>
		<value condition="String.IsEqual(ListItem.Property(stream.isvisualimpaired),true)">$LOCALIZE[39108]</value>
		<value condition="String.IsEqual(ListItem.Property(stream.isforced),true)">$LOCALIZE[39106]</value>
		<value condition="String.IsEqual(ListItem.Property(stream.isoriginal),true)">$LOCALIZE[39111]</value>
	</variable>
```

And then change the 3 separate names in Includes_DialogSelect.xml to `VideoStreamDialogItemLabelVar`

Hitcher commented:
@jjd-uk Seeing as you're working on a new dialog adding the current video stream details might be a good as well.

![screenshot00020](https://github.com/user-attachments/assets/b20bb0eb-56fa-4aac-b29b-ab8f5f325243)

Add at line 45 of Includes_SettingsDialog.xml

`<label2>[B]$INFO[VideoPlayer.VideoResolution]$INFO[Player.Process(VideoScanType)][/B]</label2>`

jjd-uk commented:
> @jjd-uk Seeing as you're working on a new dialog adding the current video stream details might be a good as well.
> 
> `<label2>[B]$INFO[VideoPlayer.VideoResolution]$INFO[Player.Process(VideoScanType)][/B]</label2>`

I already have something along those lines. It's ready but haven't had the time to push to github and create the PR, will try and get it done tomorrow.

![image](https://github.com/user-attachments/assets/2981275a-02de-4976-a7cf-c5f7f3290edd)

Hitcher commented:
> the new dialog names don't work in visible conditions, you have to use their window ID's instead.
> 
> eg `Window.IsVisible(dialogselectvideo)` does not work but `Window.IsVisible(12300)` does.

@CastagnaIT just checking if you're aware of this issue.

Thanks.

CastagnaIT commented:
i dont think can works because dialogselectvideo is a action not a window name

CastagnaIT commented:
ah wait maybe i found something, i will let you know

jjd-uk commented:
Perhaps the video version work that moved to using dialog select may give some inspiration for what is missing https://github.com/xbmc/xbmc/pull/24249

Hitcher commented:
I _think_ it's https://github.com/xbmc/xbmc/blob/master/xbmc/input/WindowTranslator.cpp but not 100%.

matthuisman commented:
Subtitle > use Interface language (Czech)
 should give the same result as
Subtitle > Czech though?

Maybe these settings always just take the part before the dash so at least
they match each other?


On Wed, 24 Sept 2025, 19:48 Stefano Gottardo, ***@***.***>
wrote:

> *CastagnaIT* left a comment (xbmc/xbmc#25762)
> <https://github.com/xbmc/xbmc/pull/25762#issuecomment-3327053729>
>
> Did you want me to raise a bug for the interface language selection issue?
>
> i think this is not a bug, but "feature" never implemented
> same problem should be seen on audio language selection(?) but atm idr
> anymore
> since kodi dont have a IETF language implementation but only workarounds
> for these
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/xbmc/xbmc/pull/25762#issuecomment-3327053729>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABPQAKPQHADMGHFE35SAVDD3UJEFLAVCNFSM6AAAAACHH55IOWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTGMRXGA2TGNZSHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

CastagnaIT commented:
from what i remember
- when you use "Interface language" setting, kodi get the value code from xml lang extension, that may use ISO639 but also can use (a kind of) IETF format, because skin can be country localized
- kodi language list name always have ISO639 format (so without dash/country), with the exception when an user have installed lang extensions or custom langs by using advanced settings, that extend the default kodi lang list, so you can 

now i understood the problem
the xml lang extension use "cs-cz"
the ISO639 is "cs" used in the default kodi languages

<img width="458" height="202" alt="immagine" src="https://github.com/user-attachments/assets/cbc7c5cc-74ff-4ed6-a958-87100fdbc843" />

its similar to `Chinese (Simple)`
<img width="857" height="350" alt="immagine" src="https://github.com/user-attachments/assets/c251a90e-0563-4af2-8f81-14fe6ae6e9bd" />
the xml lang extension has lang name of "Czech", but maybe should renamed to "Czech (Czechia)" in order to be distinguished correctly,
otherwise it will be overwrtitten (or skipped) in all language list settings
in fact if you install the czech xml lang extension on language settings appears only one entry instead of two ("Czech" and "Czech (Czechia)")
<img width="921" height="335" alt="immagine" src="https://github.com/user-attachments/assets/2fc1a8f2-08d8-424f-85c9-90c740ff55a6" />
so dont allow you to select one or the other

matthuisman commented:
ok thank you :)
Im sure this is an existing issue that I only noticed recently as i played more with the settings

Lastly, do you think the OSD -> Cog -> subtitle settings -> Subtitle stream should use your new nicer dialog select as well?
It still uses the old dialog
Same for the Audio settings -> Audio Stream

Sorry if this was asked above
It might be a skin thing

CrystalP commented:
I'm working on the addition of iso 639-3 codes and language and parsing of ietf bcp 47 tags FYI so that movie and tv shows original languages can be imported into the video db.
Then some work will be needed to improve the sub comparison function.
No ideas or plans for the GUI localization side of things, I'll leave it untouched.

axbmcuser commented:
@CastagnaIT 

Great work with this one. 👍 Easy-access to fully featured, independent and directly accessible stream selector dialogs is an awesome essential to finally have! :-)

Backported this to the most common Amlogic Kodi 21 builds while removing the automatically-enforced sorting of streams which did not feel right for a primary Kodi use case of playing local/NAS media. (where streams have an intentional authoring-order which should never be touched)
Still fully understand your thoughtful intent in regards to messy inputstream.adaptive stream lists, though.

I read that you already wrote about sorting potentially having to be followed up on at a later point. 
Played around with the implementation of a guisetting for the sorting but wasn't satisfied with it conceptionally.
(will keep it in mind & think about it some more)

I'm not too familiar with having to deal with said messy highly populated stream lists of inputstream.adaptive. 
Which add-on mostly does provide them? YouTube? Something else?

Thanks & cheers

MoojMidge commented:
> automatically-enforced sorting of streams which did not feel right for a primary Kodi use case of playing local/NAS media. (where streams have an intentional authoring-order which should never be touched)

Perhaps not the best place or time for this discussion, but I have to agree that the stream sorting doesn't seem ideal. I don't know that very many people will particularly care what the codec name of the stream is compared to resolution, framerate, bitrate, and HDR type.

> messy highly populated stream lists of inputstream.adaptive. 
> Which add-on mostly does provide them? YouTube?

At least in the case of the YouTube addon, the stream sorting is a usability regression despite the otherwise much clearer presentation of the new dialogs.

New dialogs:
<img width="896" height="523" alt="image" src="https://github.com/user-attachments/assets/7182b49b-c550-4326-b1ad-7a13ef422873" />
<img width="896" height="523" alt="image" src="https://github.com/user-attachments/assets/1ad88864-b431-4836-9ed6-5a91cbc86d97" />

Old dialogs:
<img width="896" height="523" alt="image" src="https://github.com/user-attachments/assets/5c0a8e47-b8b6-49ce-8bc9-b4b9b74d3534" />
<img width="896" height="523" alt="image" src="https://github.com/user-attachments/assets/2aefd9da-bd45-4804-bfdc-a9d5f0a4fbbf" />
